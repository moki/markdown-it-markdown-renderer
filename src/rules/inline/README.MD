# inline

this section provides implementation of the rules for the inline elements from the [commonmark spec](https://spec.commonmark.org/0.30/#inlines)

## rules list

- [Code spans](./code.ts)
- [Emphasis and strong emphasis](./emphasis.ts)
- [Autolinks](./link.ts)
- [Links](./link.ts)
- [Images](./image.ts)
- [Raw HTML](./text)
- [Textual content](./text)
- [Hard line breaks](./hardbreak.ts)
- [Soft line breaks](./softbreak.ts)
