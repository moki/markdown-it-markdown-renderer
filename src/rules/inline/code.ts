import Renderer from 'markdown-it/lib/renderer';
import Token from 'markdown-it/lib/token';

const code: Renderer.RenderRuleRecord = {
    code_inline: function (tokens: Token[], i: number) {
        const {markup, content} = tokens[i];

        return markup + content + markup;
    },
};

export {code};
export default {code};
