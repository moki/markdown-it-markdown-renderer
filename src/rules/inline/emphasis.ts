import Renderer from 'markdown-it/lib/renderer';
import Token from 'markdown-it/lib/token';

const typeToSyntax = new Map([
    ['em_open', '*'],
    ['em_close', '*'],
    ['strong_open', '**'],
    ['strong_close', '**'],
]);

function basic(tokens: Token[], i: number) {
    return typeToSyntax.get(tokens[i].type) ?? '';
}

const emphasis: Renderer.RenderRuleRecord = {
    em_open: basic,
    em_close: basic,
    strong_open: basic,
    strong_close: basic,
};

export {emphasis};
export default {emphasis};
