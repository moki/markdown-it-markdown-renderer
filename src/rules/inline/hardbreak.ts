import Renderer from 'markdown-it/lib/renderer';

const hardbreak: Renderer.RenderRuleRecord = {
    hardbreak: function () {
        return '\\\n';
    },
};

export {hardbreak};
export default {hardbreak};
