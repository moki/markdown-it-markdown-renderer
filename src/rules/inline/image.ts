import Renderer from 'markdown-it/lib/renderer';
import {MarkdownRenderer} from 'src/renderer';
import Token from 'markdown-it/lib/token';

const image: Renderer.RenderRuleRecord = {
    image: function (this: MarkdownRenderer, tokens: Token[], i: number, options, env) {
        const img = tokens[i];

        this.pending.push(img);

        const open = new Token('image_open', '', 0);

        const close = new Token('image_close', '', 0);

        const children = img.children ?? [];

        return this.renderInline([open, ...children, close], options, env);
    },
    image_open: function (this: MarkdownRenderer) {
        return '![';
    },
    image_close: function (this: MarkdownRenderer) {
        const token = this.pending.pop();
        if (token?.type !== 'image') {
            throw new Error('failed to render image token');
        }

        let rendered = '](';

        const src = token.attrGet('src');
        if (src?.length) {
            rendered += src;
        }

        const title = token.attrGet('title');
        if (title?.length) {
            if (src?.length) {
                rendered += ' ';
            }

            const markup = title.indexOf('"') === -1 ? '"' : "'";

            rendered += `${markup}${title}${markup}`;
        }

        rendered += ')';

        return rendered;
    },
};

export {image};
export default {image};
