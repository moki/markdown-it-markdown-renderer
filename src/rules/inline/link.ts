import Renderer from 'markdown-it/lib/renderer';
import {MarkdownRenderer} from 'src/renderer';
import Token from 'markdown-it/lib/token';

const link: Renderer.RenderRuleRecord = {
    link_open: function (this: MarkdownRenderer, tokens: Token[], i: number) {
        this.context.push(tokens[i]);

        return '[';
    },
    link_close: function (this: MarkdownRenderer) {
        const openToken = this.context.pop();
        if (openToken?.type !== 'link_open') {
            throw new Error('failed to render link token');
        }

        let rendered = '](';

        const href = openToken.attrGet('href');
        if (href?.length) {
            rendered += href;
        }

        const title = openToken.attrGet('title');
        if (title?.length) {
            if (href?.length) {
                rendered += ' ';
            }

            const markup = title.indexOf('"') === -1 ? '"' : "'";

            rendered += `${markup}${title}${markup}`;
        }

        rendered += ')';

        return rendered;
    },
};

export {link};
export default {link};
