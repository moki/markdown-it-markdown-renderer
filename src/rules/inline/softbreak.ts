import Renderer from 'markdown-it/lib/renderer';

const softbreak: Renderer.RenderRuleRecord = {
    softbreak: function () {
        return '\n';
    },
};

export {softbreak};
export default {softbreak};
