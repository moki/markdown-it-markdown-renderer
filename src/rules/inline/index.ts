import Renderer from 'markdown-it/lib/renderer';

import {softbreak} from './softbreak';
import {hardbreak} from './hardbreak';
import {text} from './text';
import {emphasis} from './emphasis';
import {code} from './code';
import {link} from './link';
import {image} from './image';

// https://spec.commonmark.org/0.30/#inlines
//
// html tags https://spec.commonmark.org/0.30/#raw-html
// html are handled by default html_inline and html_block rules
//
// autolinks https://spec.commonmark.org/0.30/#autolinks
// autolinks are handled by link rule
//
// images https://spec.commonmark.org/0.30/#images
//
// links
// inline links https://spec.commonmark.org/0.30/#links
// reference links: https://spec.commonmark.org/0.30/#reference-link
// link reference definition: https://spec.commonmark.org/0.30/#link-reference-definition
//
// code span `code` https://spec.commonmark.org/0.30/#code-spans
//
// emphasis
// strong - **, __
// em - _, * https://spec.commonmark.org/0.30/#emphasis-and-strong-emphasis
//
// text https://spec.commonmark.org/0.30/#textual-content
//
// hard line breaks https://spec.commonmark.org/0.30/#hard-line-breaks
//
// soft line breaks https://spec.commonmark.org/0.30/#soft-line-breaksk
const inline: Renderer.RenderRuleRecord = {
    ...code,
    ...link,
    ...image,
    ...emphasis,
    ...text,
    ...softbreak,
    ...hardbreak,
};

export {inline};
export default {inline};
