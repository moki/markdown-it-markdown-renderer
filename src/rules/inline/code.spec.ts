import {equals} from '../../../__tests__/utils';

import {MarkdownRenderer} from 'src/renderer';
import Token from 'markdown-it/lib/token';

import {code} from 'src/rules/inline/code';

describe('code_inline', () => {
    equals('inline code', '```console.log("text");```');

    it('is named code_inline', () => {
        const handler = code.code_inline;

        expect(handler).toBeTruthy();
    });

    it('return content wrapped in "`" markup for code_inline token', () => {
        const renderer = new MarkdownRenderer();

        let handler = code.code_inline;
        if (!handler) {
            return;
        }

        handler = handler.bind(renderer);

        const markup = '```';
        const content = 'console.log("text");';
        const expected = markup + content + markup;
        const tokens = [{type: 'code_inline', markup, content}];
        const actual = handler(tokens as Token[], 0, {}, {}, renderer);

        expect(actual).toBe(expected);
    });

    it('render content wrapped in "`" markup for code_inline token', () => {
        const renderer = new MarkdownRenderer({customRules: {...code}});

        const markup = '```';
        const content = 'console.log("text");';
        const expected = markup + content + markup;
        const tokens = [{type: 'code_inline', markup, content}];
        const actual = renderer.render(tokens as Token[], {}, {});

        expect(actual).toEqual(expected);
    });
});
