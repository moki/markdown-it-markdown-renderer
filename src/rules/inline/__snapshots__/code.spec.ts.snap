// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`code_inline html equals for inline code: HTML 1`] = `
"<p><code>console.log(&quot;text&quot;);</code></p>
"
`;

exports[`code_inline html equals for inline code: MD 1`] = `"\`\`\`console.log(\\"text\\");\`\`\`"`;
